java/arcs/core/analysis/Deduction.kt

```diff
@@ -0,0 +1,122 @@
+package arcs.core.analysis
+
+import arcs.core.data.expression.Expression
+
+/** Field [Identifier]. */
+typealias Identifier = String
+
+/** Lists of [Identifier]s imply an AccessPath. */
+typealias Path = List<Identifier>
+
+/**
+ * [Deduction]s represent how inputs of a Paxel [Expression] contribute to its outputs.
+ *
+ * [Deduction]s recursively associate identifiers to field paths. These relationships can be
+ * directly translated into [Claim]s and [AccessPath]s.
+ */
+sealed class Deduction {
+
+  /** Substitute aliases in all paths. */
+  abstract fun substitute(aliases: Scope): Deduction
+
+  /** Union of two [Deduction] collections. */
+  abstract operator fun plus(other: Deduction): Deduction
+
+  /** A representation of a lookup claim in a Paxel [Expression]. */
+  data class Equal(val path: Path = emptyList()) : Deduction() {
+
+    constructor(vararg paths: Identifier) : this(paths.toList())
+
+    /** Return [Equal] with all alias substitutions applied. */
+    override fun substitute(aliases: Scope): Equal = Equal(
+      path.flatMap { identifier ->
+        when (val association = aliases.associations.getOrDefault(identifier, Equal(identifier))) {
+          is Equal -> association.path
+          is Derive -> association.firstPath().path
+          is Scope -> throw UnsupportedOperationException(
+            "Cannot substitute a Scope for an Identifier."
+          )
+        }
+      }
+    )
+
+    /** Union of a [Equal] and a [Deduction]. */
+    override fun plus(other: Deduction): Deduction = when (other) {
+      is Equal -> Derive(this.path) + Derive(other.path)
+      is Derive -> Derive(this.path) + other
+      is Scope -> throw UnsupportedOperationException(
+        "Union of Equal and Scope is not well defined."
+      )
+    }
+  }
+
+  /** A representation of a Derivation claim in a Paxel [Expression]. */
+  data class Derive(val paths: List<Path> = emptyList()) : Deduction() {
+
+    constructor(vararg paths: Path) : this(paths.toList())
+
+    /** Union of a [Derive] and another [Deduction]. */
+    override fun plus(other: Deduction): Deduction = when (other) {
+      is Equal -> this + Derive(other.path)
+      is Derive -> Derive(paths + other.paths)
```

Comment:
Something to check: `x + x` should be `Derive(listOf(x))`, not `Derive(listOf(x), listOf(x))`